app/models/concerns/course/levels_concern.rb

```diff
@@ -41,4 +41,35 @@ def compute_level_number(experience_points)
     level = compute_level(experience_points)
     level ? level.level_number : 0
   end
+
+  # Returns the current level's experience point (exp) threshold for
+  # for a given number of experience_points.
```

Comment:
How about: retaining the nil next level, but ensuring that the current course user always returns a valid level?

I *think* that the edge cases should be completely isolated within the level model. That might make the messiness confined to one class, instead of having implementation details splashed all over the code base.